torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,157 @@
+from dataclasses import dataclass
+from typing import Dict, Tuple
+
+import torch
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+import torchao.prototype.autoround.utils as ar_utils
+import torchao.quantization as ao_quant
+from torchao.dtypes import to_affine_quantized_static
+from torchao.prototype.autoround.multi_tensor import MultiTensor
+
+# TODO: remove it before merge
+ar_utils.freeze_random()
+
+
+@dataclass
+class AutoRoundConfig:
+    bits: int = 4
+    sym: bool = False
+    iters: int = 200
+    group_size: int = 128
+    train_bs: int = 4
+    eval_bs: int = 4
+    seed: int = 42
+    amp: bool = True
+    nsamples: int = 128
+    seqlen: int = 2048
+    quant_lm_head: bool = False
+
+
+auto_round_config = AutoRoundConfig()
+
+
+@torch.no_grad()
+def create_qmodel_from_qdq_model(qdq_model: torch.nn.Module):
+    """Create a quantized model from a qdq model.
+
+    The qdq_model includes Linear quantized by auto-round, which includes qdq weight, scale, zp.
+    """
+
+    @torch.no_grad()
+    def apply_static_quant(observed_linear):
+        device = observed_linear.weight.device
+        weight_scale = observed_linear.scale.to(device)
+        weight_zero_point = observed_linear.zp.to(device)
+
+        def weight_quant_func(weight):
+            block_size = (1, observed_linear.group_size)
+            # TODO: shift the zero and prepack the weight to use tinygemm?
+            return to_affine_quantized_static(
+                input_float=weight,
+                scale=weight_scale,
+                zero_point=weight_zero_point,
+                block_size=block_size,
+                target_dtype=torch.uint8,
+                quant_min=0,
+                quant_max=15,
+                zero_point_domain=ao_quant.quant_primitives.ZeroPointDomain.INT,
```

Comment:
are you trying to use int4 tinygemm kernel? if so you'll need to follow this setup: https://github.com/pytorch/ao/blob/174e630af2be8cd18bc47c5e530765a82e97f45b/torchao/quantization/quant_api.py#L402-L412